apps/src/dance/blockly/buildDanceBlockly.ts

```diff
@@ -8,17 +8,34 @@ import {
 
 import getBlockOptions from './getBlockOptions';
 
-const SET_BACKGROUND = 'Dancelab_setBackgroundEffectWithPaletteAI';
-const SET_FOREGROUND = 'Dancelab_setForegroundEffectExtended';
-const CHANGE_MOVE = 'Dancelab_changeMoveEachLR';
+const DANCELAB_PREFIX = 'Dancelab_';
+const GENERATED_PREFIX = 'GeneratedDancers_';
 
-function randomElement<T>(array: readonly T[]): T {
-  return array[Math.floor(Math.random() * array.length)];
+// Try to find a block definition by type string.
+// We don't rely on the exact shape of BlockDefinition; cover common keys.
+function hasBlockType(defs: BlockDefinition[], name: string): boolean {
+  return defs.some(d => d.name === name);
 }
 
-function uidFactory(prefix: string): () => string {
-  let i = 0;
-  return () => `${prefix}_${++i}`;
+// Given a block type (without block prefix), get the preferred, defined type.
+// Prefers blocks from the GeneratedDancers pool over the standard Dancelab pool.
+// Block pools may vary by level, but all Dance level use the Dancelab pool.
+function getPreferredBlockType(
```

Comment:
Does this mean we might have `GeneratedDancer_` variants of setBackground/setForeground?

Reply:
No, not at this time. However, it does mean that if they were to be added at any point, including by levelbuilder they'd work immediately.
I expect we'll be iterating on the generated code a fair bit in the coming weeks, including more new blocks, so this utility was aiming to be as flexible as possible.